### PR TITLE
Planutils-driven routing

### PR DESCRIPTION
**Pull request by haz**

### CODE REVIEW

**Comment by camcunningham** [comment]
@haz, Let's iterate on this and get it up to snuff when it comes to executing generic packages - would like to get some extra eyes on it to see where we can fix / add some of the missing functionality that you want.

File: server/celery-queue/tasks.py [hunk 1]
```diff
@@ -17,10 +20,31 @@ def download_file( url: str, dst: str):
     r = requests.get(url)
     with open(dst, 'wb') as f:
         f.write(r.content)
+        
+def retrieve_output_file(target_file:str, folder):
+    available_files = os.listdir(folder)
```
Comment (haz) [inline server/celery-queue/tasks.py:25]:
```python
with open(os.path.join(folder, target_file['file']), 'r') as f:
    output = f.read()
if target_file['type'] == 'json':
    output = json.loads(output)
return output
```

File: server/celery-queue/tasks.py [hunk 2]
```diff
@@ -32,20 +56,66 @@ def solve(domain_url: str, problem_url: str, solver: str) -> str:
...
+    res = subprocess.run(call, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                        executable='/bin/bash', encoding='utf-8',
+                        shell=True, cwd=tmpfolder)
+    
+    output = retrieve_output_file(output_file, tmpfolder)
+
+    return {"stdout":res.stdout, "stderr":res.stderr, "call":call, "output":output}
+    
+# Running a generic solver with string PDDL inputs 
+@celery.task(name='tasks.solve.string')
```
Comment (haz) [inline server/celery-queue/tasks.py:101]: Just like the suggestion above, can you fold this into the previous by just having the right details/defaults in advance?

**Review by haz — COMMENTED** [review]
Looking pretty good! Just some minor things. Let me know if any of it doesn't make sense.

File: server/api/app.py [hunk 3]
```diff
@@ -73,15 +74,71 @@ def index():
...
+    # Get request
+    if request.method == 'GET':
+        # This is where we will send the user to the API documentation
+        return jsonify({PACKAGES[package]})
+    
+    # Post request
+    elif request.method == 'POST':
+        # Called route with a package that isn't in Planutils
+        if package not in PACKAGES:
+            return make_response("That package does not exist.", 400)
```
[the remark this answers is not in the document]
Comment (camcunningham) [reply server/api/app.py:89]: Could you clarify this? Do you want the make_response to be replaced with jsonify?

[the remark this answers is not in the document]
Comment (haz) [reply server/api/app.py:89]: Aye -- I thought `make_response` would create an HTML or plain/text or whatever response. This is a `POST` request (line 86), so the error should be in a json format.

File: server/api/app.py [hunk 4]
```diff
@@ -73,15 +74,71 @@ def index():
         # os.remove( pddl_files.path(filename_problem) )
 
         return redirect(url_for('index'))
+    
+@app.route('/package/<package>/<service>', methods=['GET', 'POST'])
+def runPackage(package, service):
+    # Get request
+    if request.method == 'GET':
+        # This is where we will send the user to the API documentation
+        return jsonify({PACKAGES[package]})
```
[the remark this answers is not in the document]
Comment (camcunningham) [reply server/api/app.py:83]: Added in a check, let me know if it looks good

File: server/celery-queue/tasks.py [hunk 5]
```diff
@@ -17,10 +20,31 @@ def download_file( url: str, dst: str):
...
+            if target_file['type'] == 'json':
+                output[file] = json.loads(f.read())
+            else:
+                output[file] = f.read()
+            break
+    return output
+
+def write_to_temp_file(name:str, data:str, folder:str):
+    path = os.path.join(folder, name)
+    file = open(path, "w")
```
[the remark this answers is not in the document]
Comment (camcunningham) [reply server/celery-queue/tasks.py:40]: Replaced

File: server/celery-queue/tasks.py (hunk 1 above)
Comment (camcunningham) [reply server/celery-queue/tasks.py:25]: Replaced

File: server/celery-queue/tasks.py (hunk 2 above)
Comment (camcunningham) [reply server/celery-queue/tasks.py:101]: Going to get this done today

Comment (camcunningham) [reply server/celery-queue/tasks.py:101]: Do you think that we should combine tasks.solve and tasks.solve.string, with the separate task for running generic packages - or should I look to combine all 3 into tasks.run.package?

@haz

Comment (haz) [reply server/celery-queue/tasks.py:101]: Ultimately it should be just the single package handling. We'll want to handle the legacy `/solve` endpoint for a while, but just repackage the data into a call for the `tasks.run.package`

File: server/api/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (haz) [reply server/api/app.py:83]: Need to push still, I think. Feel free to resolve those comments you've already tackled.

[the remark this answers is not in the document]
Comment (camcunningham) [reply server/api/app.py:83]: Should be up soon, just doing all the fixes in one commit for now

File: server/celery-queue/tasks.py (hunk 2 above)
Comment (camcunningham) [reply server/celery-queue/tasks.py:101]: Sounds good

**Comment by camcunningham** [comment]
@haz I've pushed the changes

**Comment by haz** [comment]
Do we assume that every argument in the call string has an argument filled in? Any way to check this? For optional args, we may want to allow empty filling.

**Pull request merged**